### PR TITLE
Add validation of trace request headers

### PR DESCRIPTION
**Pull request by twometresteve**
## Goal

Adds validation of the following headers on every trace request received
- Bugsnag-Api-Key – 32 hex char guid
- Bugsnag-Sent-At – ISO 8601 date
- Bugsnag-Span-Sampling – of the format x:y where
  - x is a decimal between 0 and 1 (inclusive)
  - y is an integer 

## Design

Existing mechanism extended to include steps for validating headers.

## Tests

Unit and e2e tests updated and I've also run the whole set of Android and Cocoa performance e2e tests against it.

### CODE REVIEW

File: lib/maze/schemas/trace_validator.rb [hunk 1]
```diff
@@ -5,6 +5,10 @@
 module Maze
   module Schemas
 
+    HEX_STRING_16 = '^[A-Fa-f0-9]{16}$'
+    HEX_STRING_32 = '^[A-Fa-f0-9]{32}$'
+    SAMPLING_HEADER = '^(1(.0)?|0(\.[0-9]+)?):[0-9]+$'
```
Comment (imjoehaines) [inline lib/maze/schemas/trace_validator.rb:10]: If there's more than 1 sampling probability then they are all listed in the header separated by semicolons, i.e. `a:b;c:d`, `a:b;c:d;e:f` etc&hellip;

Comment (twometresteve) [reply lib/maze/schemas/trace_validator.rb]: Regex updated and new test cases added.

File: lib/maze/schemas/trace_validator.rb [hunk 2]
```diff
@@ -40,6 +45,61 @@ def validate
         )
       end
 
+      def check(date)
```
Comment (Cawllec) [inline lib/maze/schemas/trace_validator.rb:48]: This doesn't appear to be used, and has an ambiguous name, could it be removed or renamed?

File: lib/maze/schemas/trace_validator.rb [hunk 3]
```diff
@@ -40,6 +45,61 @@ def validate
...
+          expected = Regexp.new(HEX_STRING_32)
+          unless expected.match(api_key)
+            @success = false
+            @errors << "bugsnag-api-key header was expected to match the regex '#{HEX_STRING_32}', but was '#{api_key}'"
+          end
+        end
+
+        # Bugsnag-Sent-at
+        validate_header('bugsnag-sent-at') do |date|
+          begin
```
Comment (Cawllec) [inline lib/maze/schemas/trace_validator.rb:75]: Can this use the method below, `def validate_date(date)`?

File: lib/maze/schemas/trace_validator.rb [hunk 4]
```diff
@@ -15,19 +20,19 @@ class TraceValidator
...
       #
-      #   @param body [Hash] The body of the trace to validate
-      def initialize(body)
-        @body = body
-        @success = nil
+      #   @param request [Hash] The trace request to validate
+      def initialize(request)
+        @headers = request[:request].header
+        @body = request[:body]
+        @success = true
```
Comment (Cawllec) [inline lib/maze/schemas/trace_validator.rb:27]: Could this introduce false positives if the validation isn't actually run?

**Review by Cawllec — APPROVED** [review]
Pending tests passing

**Pull request merged**